Core/clim-basic/ports.lisp

```diff
@@ -171,107 +176,183 @@
+             (setf (port-pointer-sheet port) nil)
+             (dispatch-event-copy sheet event 'pointer-exit-event))
+           (map-within-mirror (function sheet x y)
+             (funcall function sheet)
+             (map-over-sheets-containing-position
+              (lambda (child)
+                (unless (sheet-direct-mirror child)
+                  (multiple-value-bind (child-x child-y)
+                      (untransform-position (sheet-transformation child) x y)
+                    (map-within-mirror function child child-x child-y))))
+              sheet x y)))
+    ;; FIXME clx generates uneven number of ungrab-enter/ungrab-leave
+    ;; events what may disrupt the enter/exit event balance for a
+    ;; sheet. Not distributing such events makes overall UX worse, for
+    ;; instance user needs first to click on a window to press an
+    ;; armed push-button. What's the right thing? -- jd 2019-08-28
+    (typecase event
+      ;; (pointer-ungrab-enter-event (warn "ungrab enter event"))
+      ;; (pointer-ungrab-leave-event (warn "ungrab leave event"))
+      (pointer-enter-event
+       (let ((sheet (event-sheet event)))
+         (get-pointer-position (sheet event)
+           (map-within-mirror #'synthetize-enter sheet x y))))
+      (pointer-exit-event
+       (when-let ((sheet (event-sheet event))
+                  (child (port-pointer-sheet port)))
+         (when (eql sheet (sheet-mirrored-ancestor child))
+           (loop
+              for over-sheet = child then (sheet-parent over-sheet)
+              do (synthetize-exit over-sheet)
+              until (eql over-sheet sheet)))))
+      (otherwise
+       (flet ((common-ancestor (sheet-a sheet-b)
+                (loop
+                   (cond ((or (null sheet-a) (null sheet-b)
+                              (sheet-direct-mirror sheet-a)
+                              (sheet-direct-mirror sheet-b))
+                          (return-from common-ancestor nil))
+                         ((sheet-ancestor-p sheet-b sheet-a)
+                          (return-from common-ancestor sheet-a))
+                         (t (setf sheet-a (sheet-parent sheet-a)))))))
+         (let* ((old-sheet (port-pointer-sheet port))
+                (new-sheet (compute-pointer-event-sheet event))
+                (common-ancestor (common-ancestor old-sheet new-sheet)))
+           ;; distribute exit events (innermost first)
+           (do ((sheet old-sheet (sheet-parent sheet)))
+               ((or (null sheet)
+                    (eq sheet common-ancestor)
+                    (sheet-direct-mirror sheet)))
+             (synthetize-exit sheet))
+           ;; distribute enter events (innermost last)
+           (do ((sheet new-sheet (sheet-parent sheet))
+                (sheets nil))
+               ((or (null sheet)
+                    (eq sheet common-ancestor)
+                    (sheet-direct-mirror sheet))
+                (dolist (sheet sheets)
```

Comment:
I would much prefer `(loop with sheets … finally (map nil #'synthetize-enter sheets))`.

Reply:
for me do is much clearer here.